sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetColumnVector.java

```diff
@@ -0,0 +1,379 @@
+   */
+  WritableColumnVector getRepetitionLevelVector() {
+    return this.repetitionLevels;
+  }
+
+  /**
+   * Returns the writable column vector used to store definition levels.
+   */
+  WritableColumnVector getDefinitionLevelVector() {
+    return this.definitionLevels;
+  }
+
+  /**
+   * Returns the column reader for reading a Parquet column.
+   */
+  VectorizedColumnReader getColumnReader() {
+    return this.columnReader;
+  }
+
+  /**
+   * Sets the column vector to 'reader'. Note this can only be called on a primitive Parquet
+   * column.
+   */
+  void setColumnReader(VectorizedColumnReader reader) {
+    if (!isPrimitive) {
+      throw new IllegalStateException("Can't set reader for non-primitive column");
+    }
+    this.columnReader = reader;
+  }
+
+  private void calculateCollectionOffsets() {
+    int maxDefinitionLevel = column.definitionLevel();
+    int maxElementRepetitionLevel = column.repetitionLevel();
+
+    ParquetColumn child = column.children().apply(0);
+    int maxElementDefinitionLevel = child.definitionLevel();
+    if (!child.required()) maxElementDefinitionLevel -= 1;
+
+    // There are 4 cases when calculating definition levels:
+    //   1. definitionLevel == maxDefinitionLevel
+    //     ==> value is defined and not null
+    //   2. definitionLevel == maxDefinitionLevel - 1
+    //     ==> value is null
+    //   3. definitionLevel < maxDefinitionLevel - 1
+    //     ==> value doesn't exist since one of its optional parents is null
+    //   4. definitionLevel > maxDefinitionLevel
+    //     ==> value is a nested element within an array or map
+    //
+    // `i` is the index over all leaf elements of this array, while `offset` is the index over
+    // all top-level elements of this array.
+    int rowId = 0;
+    for (int i = 0, offset = 0; i < definitionLevels.getElementsAppended();
+         i = getNextCollectionStart(maxElementRepetitionLevel, i)) {
+      vector.reserve(rowId + 1);
+      int definitionLevel = definitionLevels.getInt(i);
+      if (definitionLevel < maxElementDefinitionLevel) {
+        // This means the collection is null, and it is not an array element. In this case, we
+        // should increase offset to skip it when returning an array starting from the offset.
+        //
+        // For instance, considering an array of strings with 2 elements like the following:
+        //  null, [a, b, c]
+        // the child array (which is of String type) in this case will be:
+        //  null:   1 0 0 0
+        //  length: 0 1 1 1
+        //  offset: 0 0 1 2
+        // and the array itself will be:
+        //  null:   1 0
+        //  length: 0 3
+        //  offset: 0 1
+        //
+        // It's important that for the second element `[a, b, c]`, the offset starts from 1
+        // since otherwise we'd include the first null element from child array in the result.
+        offset += 1;
+      }
+      if (definitionLevel <= maxDefinitionLevel - 1) {
```

Comment:
why this is not `else if`? Could two cases happen at the same time?

Reply:
Yes we can merge them (just need to add `offset += 1` in the first two if branches. I separated it out mainly to add the block of comments explaining why we need to increment `offset`.